drivers: pwm: refactor esp32 pwm ledc driver

LucasTambor opened the pull request:
Use HAL for low level access.
Use device tree for hardware configuration.
Support for esp32s2 and esp32c3.

LucasTambor committed:
dts: esp32: full ledc configuration in binding

This commit moves the hardware configuration for ledc
peripheral to the device-tree instead of Kconfig.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
drivers: pwm: refactor esp32 pwm ledc driver

Use hal function for low level access.
Use device tree for hardware configuration.
Support for esp32s2.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
doc: boards: add pwm to espressif's boards

Add PWM support to espressif's boards documentation

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
west.yml: hal_espressif: updates to latest revision

Includes LEDC pinctrl support and source files.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

mbolivar-nordic reviewed (commented):
@anangl I am trying to offload subsystem-specific devicetree work onto the maintainers of those subsystems, so I likely will not be reviewing the DT portions of this since I consider your devicetree skills are expert.

The reason why I am doing this is that I cannot possibly review all devicetree changes in all subsystems, and need to rely on maintainers' skills for the project to continue to scale.

Please feel free to ping me here or offline if you have any particular questions though!

mbolivar-nordic commented:
> @anangl I am trying to offload subsystem-specific devicetree work onto the maintainers of those subsystems, so I likely will not be reviewing the DT portions of this since I consider your devicetree skills are expert.

@henrikbrixandersen @gmarull FYI as well since you are collaborators on PWM

gmarull commented on boards/xtensa/esp_wrover_kit/esp_wrover_kit.dts:
```diff
@@ -111,6 +133,29 @@
+	channel0@0 {
+		reg = <0x0>;
+		timer = <0>;
+		freq = <100>;
+	};
+	channel1@1 {
+		reg = <0x1>;
+		timer = <1>;
+		freq = <100>;
+	};
+	channel2@2 {
+		reg = <0x2>;
+		timer = <2>;
+		freq = <100>;
```
this seems to duplicate the frequency in the pwm cells, am I wrong?

gmarull commented on dts/bindings/pwm/espressif,esp32-ledc.yaml:
```diff
@@ -1,15 +1,127 @@
+    const: 3
+
+  pinctrl-0:
+    type: phandles
```
please don't duplicate type here, inherit from `pinctrl-device.yaml`

gmarull commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -1,340 +1,303 @@
-	PWM_LED_ESP32_REF_TICK_FREQ,
-	PWM_LED_ESP32_APB_CLK_FREQ,
-};
+#define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pwm_ledc_esp32);
```
`LOG_MODULE_REGISTER(pwm_ledc_esp32, CONFIG_PWM_LOG_LEVEL);`

gmarull commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -344,135 +307,40 @@ static const struct pwm_driver_api pwm_led_esp32_api = {
-/* Initialization for PWM_LED_ESP32 */
-#include <zephyr/device.h>
-#include <zephyr/init.h>
+#define CHANNEL_CONFIG(node_id)								\
+	{	.idx = DT_REG_ADDR(node_id),						\
+		.channel_num = DT_REG_ADDR(node_id) % 8,			\
+		.timer_num = DT_PROP(node_id, timer),				\
+		.freq = DT_PROP(node_id, freq),						\
+		.speed_mode = DT_REG_ADDR(node_id) < SOC_LEDC_CHANNEL_NUM ?	\
+			LEDC_LOW_SPEED_MODE : !LEDC_LOW_SPEED_MODE,		\
+		.clock_src = LEDC_APB_CLK,							\
+	},
```
please align `\` to the same column (or run `clang-format` ;-) )

gmarull reviewed (changes requested):
thanks for this cleanup, please check my comments (minor issues)

gmarull commented on boards/xtensa/esp_wrover_kit/esp_wrover_kit.dts:
```diff
@@ -46,6 +52,22 @@
+
+	pwmleds {
+		compatible = "pwm-leds";
+		pwm_led_red: pwm_led_0 {
+		   label = "Red PWM LED";
+		   pwms = <&ledc0 0 PWM_HZ(100) PWM_POLARITY_NORMAL>;
+		};
+		pwm_led_green: pwm_led_1 {
+		   label = "Green PWM LED";
+		   pwms = <&ledc0 1 PWM_HZ(100) PWM_POLARITY_NORMAL>;
+		};
+		pwm_led_blue: pwm_led_2 {
+		   label = "Blue PWM LED";
+		   pwms = <&ledc0 2 PWM_HZ(100) PWM_POLARITY_NORMAL>;
```
mixes tabs/spaces

LucasTambor replied:
Thanks!

gmarull commented on dts/bindings/pwm/espressif,esp32-ledc.yaml:
```diff
@@ -1,15 +1,114 @@
+  It is possible to include multiple channel's mapping in the 'pinmux' property:
+
+    &ledc0 {
+      pinctrl-0 = <&ledc0_ch0_gpio0 &ledc0_ch9_gpio2 &ledc0_ch10_gpio4>;
+    }
+
+  Use the child bindings to configure the desired channel:
+
+    &ledc0 {
+      pinctrl-0 = <&ledc0_ch0_gpio0 &ledc0_ch9_gpio2 &ledc0_ch10_gpio4>;
```
this example is missing `pinctrl-names`

LucasTambor replied:
Thanks!

gmarull commented on dts/bindings/pwm/espressif,esp32-ledc.yaml:
```diff
@@ -1,15 +1,114 @@
+  The mapping between the channel and GPIO is done through pinctrl
+
+    &ledc0 {
+      pinctrl-0 = <&ledc0_ch0_gpio0>;
```
ditto

gmarull commented on samples/basic/blinky_pwm/boards/esp32.overlay:
```diff
@@ -2,25 +2,42 @@
+     aliases {
+         pwm-0 = &ledc0;
+         pwm-led0 = &pwm_led_blue;
+     };
 
-	pwmleds {
-		compatible = "pwm-leds";
-		pwm_led_blue: pwm_led_gpio0_2 {
-		   label = "PWM LED0";
-		   pwms = <&ledc0 2 1000 PWM_POLARITY_NORMAL>;
-		};
-	};
+     pwmleds {
+         compatible = "pwm-leds";
+         pwm_led_blue: pwm_led_gpio0_2 {
+            label = "PWM LED0";
+            pwms = <&ledc0 0 1000 PWM_POLARITY_NORMAL>;
+         };
+     };
+ };
+
+&pinctrl {
+    ledc0_ch0_gpio2: ledc0_ch0_gpio2 {
+        pinmux = <LEDC_CH0_GPIO2>;
+        output-enable;
+    };
 };
```
please use tabs, check your editor settings

gmarull replied:
also, why isn't this defined in the board dts?

LucasTambor replied:
1. Thanks for the tip. 
2. The esp32 is based on the esp32_devkitc board, wich has no built-in LED. Since the board itself has no native support for the LED I prefer not to define it in the dts, but to include this overlay so the regular user can run the sample with an external LED. I will include the board in the README of this sample with instructions to connect an external LED.

gmarull commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -1,478 +1,351 @@
+
+static struct pwm_ledc_esp32_data pwm_ledc_esp32_data = {
+	.hal = {
+		.dev = (ledc_dev_t *) DT_REG_ADDR(DT_NODELABEL(ledc0)),
```
why isn't this using `_INST` macros?

LucasTambor replied:
Do you mean like .dev = (ledc_dev_t *) DT_REG_ADDR(DT_INST(0, espressif_esp32_ledc)) ?

LucasTambor committed:
samples: basic: blinky_pwm: add support for espressif boards

Add overlay for espressif boards that does not have a built-in LED.

Add instructions on README file about the required LED connection.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

sylvioalves reviewed (approved):
Sounds good for me. It also adds an overlay file so that `blinky_pwm`can be used.
